### PR TITLE
YONK-791: Mobile app Detail & List APIs with Unit Tests

### PR DESCRIPTION
**Pull request by aamishbaloch**
The PR include the initial setup of the mobile app with model details. List and details API are added along with the tests.

### CODE REVIEW

**Comment by aamishbaloch** [comment]
@ziafazal Can you review this?

File: mobileapps/tests.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+                'name': 'Test Mobile App {}'.format(i),
+                'identifier': 'Test Mobile App Identifier {}'.format(i),
+                'current_version': '1.0',
+                'updated_by': self.test_mobileapp_updated_by,
+                'users': [user.id for user in users],
+                'organizations': [organization.id for organization in organizations],
+            }
+            mobile_apps.append(self.setup_test_mobileapp(mobileapp_data=data))
+
+        test_uri = '{}'.format(self.base_mobileapps_uri)
```
Comment (ziafazal) [inline mobileapps/tests.py]: I think we should use `reverse` here and all other places to resolve the urls.

File: mobileapps/tests.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+@ddt.ddt
+class MobileappsApiTests(ModuleStoreTestCase, APIClientMixin):
+    """ Test suite for Mobileapps API views """
+
+    MODULESTORE = TEST_DATA_SPLIT_MODULESTORE
+
+    def setUp(self):
+        super(MobileappsApiTests, self).setUp()
+        self.test_server_prefix = 'https://testserver'
```
Comment (ziafazal) [inline mobileapps/tests.py]: these can be eliminated if we use `reverse` on relevant views to resolve urls.

File: mobileapps/tests.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+        self.base_mobileapps_uri = '/api/server/mobileapps/'
+
+        self.test_user_email = str(uuid.uuid4())
+        self.test_user_username = str(uuid.uuid4())
+
+        self.test_user = User.objects.create(
```
Comment (ziafazal) [inline mobileapps/tests.py]: Why not using `UserFactory`?

File: mobileapps/tests.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    MODULESTORE = TEST_DATA_SPLIT_MODULESTORE
+
+    def setUp(self):
+        super(MobileappsApiTests, self).setUp()
+        self.test_server_prefix = 'https://testserver'
+        self.base_mobileapps_uri = '/api/server/mobileapps/'
+
+        self.test_user_email = str(uuid.uuid4())
+        self.test_user_username = str(uuid.uuid4())
```
Comment (ziafazal) [inline mobileapps/tests.py]: It appears we don't need these username and email properties.

File: mobileapps/tests.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+        self.test_user = User.objects.create(
+            email=self.test_user_email,
+            username=self.test_user_username
+        )
+
+        self.test_mobileapp_name = str(uuid.uuid4())
+        self.test_mobileapp_identifier = "dsfcsdfc"
```
Comment (ziafazal) [inline mobileapps/tests.py]: can we use UUID for identifier here too?

File: mobileapps/tests.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data['count'], 1)
+        self.assertEqual(response.data['results'][0]['identifier'], "ABC")
+
+        response = self.do_get('{}?organization_name=query'.format(test_uri))
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.data['count'], 0)
+
+    def test_mobileapps_detail_get(self):
```
Comment (ziafazal) [inline mobileapps/tests.py:170]: we need to add test for put, patch and delete methods.

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch complete first round of review.

**Comment by aamishbaloch** [comment]
@ziafazal Done with the suggested changes. Can you review the changes?

File: mobileapps/tests.py [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+                'name': 'Test Mobile App {}'.format(i),
+                'identifier': 'Test Mobile App Identifier {}'.format(i),
+                'current_version': '1.0',
+                'updated_by': self.test_mobileapp_updated_by,
+                'users': [user.id for user in users],
+                'organizations': [organization.id for organization in organizations],
+            }
+            mobile_apps.append(self.setup_test_mobileapp(mobileapp_data=data))
+
+        test_uri = '{}'.format(self.base_mobileapps_uri)
```
Comment (ziafazal) [inline mobileapps/tests.py]: `test_uri` does not make sense here. I think we should use `self.base_mobileapps_uri`

File: mobileapps/tests.py [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+        self.assertEqual(len(response.data), len(mobile_apps))
+
+    def test_mobileapps_list_search_by_app_name(self):
+        mobile_apps = []
+
+        mobile_apps.append(self.setup_test_mobileapp(mobileapp_data={"name": 'ABC App', "identifier": "ABC"}))
+        mobile_apps.append(self.setup_test_mobileapp(mobileapp_data={"name": 'LMN App', "identifier": "LMN"}))
+        mobile_apps.append(self.setup_test_mobileapp(mobileapp_data={"name": 'XYZ App', "identifier": "XYZ"}))
+
+        test_uri = '{}'.format(self.base_mobileapps_uri)
```
Comment (ziafazal) [inline mobileapps/tests.py]: `test_uri` not needed.

File: mobileapps/tests.py [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+        self.assertIn("provider_key", response.data)
+        self.assertIn("provider_secret", response.data)
+        self.assertIn("provider_dashboard_url", response.data)
+        self.assertIn("current_version", response.data)
+        self.assertIsNotNone(response.data['created'])
+        self.assertIsNotNone(response.data['modified'])
+
+    def test_mobileapps_detail_put(self):
+
+        mobileapp = MobileApp.objects.create(
```
Comment (ziafazal) [inline mobileapps/tests.py:194]: why not using  `self.setup_test_mobileapp(mobileapp_data=data)` method to create mobiel app?

File: mobileapps/tests.py [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+            name='ABC App',
+            identifier='ABC',
+            operating_system=self.test_mobileapp_operating_system,
+            current_version=self.test_mobileapp_current_version,
+            updated_by=self.test_user,
```
Comment (ziafazal) [inline mobileapps/tests.py]: `updated_by` should be pulling from the `request.user` not from the data.

File: mobileapps/tests.py [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+        data = {
+            'name': "XYZ App",
+            'identifier': mobileapp.identifier,
+            'operating_system': mobileapp.operating_system,
+            'current_version': mobileapp.current_version,
+            'updated_by': mobileapp.updated_by.id,
+        }
+
+        response = self.do_put(reverse('mobileapps-detail', kwargs={'pk': mobileapp.id}), data=data)
+        self.assertEqual(response.status_code, 200)
```
Comment (ziafazal) [inline mobileapps/tests.py]: we should assert here `put` is updating complete resource.

File: mobileapps/tests.py [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+            current_version=self.test_mobileapp_current_version,
+            updated_by=self.test_user,
+        )
+
+        data = {
+            'name': "XYZ App",
+        }
+
+        response = self.do_patch(reverse('mobileapps-detail', kwargs={'pk': mobileapp.id}), data=data)
+        self.assertEqual(response.status_code, 200)
```
Comment (ziafazal) [inline mobileapps/tests.py:231]: we should assert hare `patch` is updating resource partially.

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch completed second round of review.

File: mobileapps/tests.py (hunk 9 above)
Comment (aamishbaloch) [reply mobileapps/tests.py:194]: I am using this just to ease myself by getting the object not response. It is easier to get values from object which I have to use after creation.

**Comment by aamishbaloch** [comment]
@ziafazal Can you review the changes?

File: mobileapps/models.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+    provider_secret = models.CharField(max_length=255, null=True, blank=True)
+    provider_dashboard_url = models.CharField(max_length=255, null=True, blank=True)
+    current_version = models.CharField(max_length=255)
+    is_active = models.BooleanField(default=True)
+    updated_by = models.ForeignKey(User, on_delete=models.PROTECT, blank=True)
+    users = models.ManyToManyField(User, related_name="mobile_apps", blank=True)
+    organizations = models.ManyToManyField(Organization, related_name="mobile_apps", blank=True)
+
+    @property
+    def get_provider_key(self):
```
Comment (ziafazal) [inline mobileapps/models.py]: `get` prefix should be for methods. Can we rename this to  `provider_key_decrypted`? same for `provider_secret`.

**Review by ziafazal — CHANGES_REQUESTED** [review]
@aamishbaloch a couple of minor changes other than these LGTM.

**Review by ziafazal — APPROVED** [review]
👍

**Pull request merged**